src/main/java/com/google/cloud/tools/appengine/cloudsdk/process/LegacyProcessHandler.java

```diff
@@ -0,0 +1,267 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.appengine.cloudsdk.process;
+
+import com.google.cloud.tools.appengine.api.AppEngineException;
+import com.google.cloud.tools.appengine.cloudsdk.internal.process.WaitingProcessOutputLineListener;
+import com.google.common.base.Charsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+/** Process handler that mimics the previous behavior of ProcessRunner. */
+public class LegacyProcessHandler implements ProcessHandler {
+  private final List<ProcessOutputLineListener> stdOutLineListeners;
+  private final List<ProcessOutputLineListener> stdErrLineListeners;
+  private final List<ProcessExitListener> exitListeners;
+  private final List<ProcessStartListener> startListeners;
+  private final WaitingProcessOutputLineListener waitingProcessOutputLineListener;
+  private final boolean async;
+
+  // TODO: historically it looks like this code hasn't been testable, we need to pass an
+  // TODO: executor service and get rid of the separate thread instantiation that we've done here.
+
+  /**
+   * Non-public constructor, but waitingProcessOutputLineListener must be part of the other
+   * listeners (stdout, stderr, processExit) to be triggered correctly.
+   */
+  LegacyProcessHandler(
+      boolean async,
+      List<ProcessOutputLineListener> stdOutLineListeners,
+      List<ProcessOutputLineListener> stdErrLineListeners,
+      List<ProcessStartListener> processStartListeners,
+      List<ProcessExitListener> processExitListeners,
+      WaitingProcessOutputLineListener waitingProcessOutputLineListener) {
+    this.async = async;
+    this.stdOutLineListeners = stdOutLineListeners;
+    this.stdErrLineListeners = stdErrLineListeners;
+    this.exitListeners = processExitListeners;
+    this.startListeners = processStartListeners;
+    this.waitingProcessOutputLineListener = waitingProcessOutputLineListener;
+  }
+
+  @Override
+  public void handleProcess(Process process) throws ProcessHandlerException {
+    Thread stdOutHandler = null;
+    Thread stdErrHandler = null;
+    try {
+
+      // Only handle stdout or stderr if there are listeners.
+      if (!stdOutLineListeners.isEmpty()) {
+        stdOutHandler = handleStdOut(process);
+      }
+      if (!stdErrLineListeners.isEmpty()) {
+        stdErrHandler = handleErrOut(process);
+      }
+
+      for (ProcessStartListener startListener : startListeners) {
+        startListener.onStart(process);
+      }
+
+      if (async) {
+        asyncRun(process, stdOutHandler, stdErrHandler);
+      } else {
+        shutdownProcessHook(process);
+        syncRun(process, stdOutHandler, stdErrHandler);
+      }
+
+    } catch (InterruptedException | AppEngineException ex) {
+      throw new ProcessHandlerException(ex);
+    }
+  }
+
+  private Thread handleStdOut(Process process) {
```

Comment:
You can factor out the common code in these two methods by passing in an input stream from the process.